File: src/Convert_Sorted_Array_to_Binary_Search_Tree.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def build(self, l, r):
        mid = (l + r) >> 1
        node = TreeNode(self.nums[mid])
        if mid - l:
            node.left = self.build(l, mid)
        if r - mid - 1:
            node.right = self.build(mid + 1, r)
        return node

    # @param {integer[]} nums
    # @return {TreeNode}
    def sortedArrayToBST(self, nums):
        if nums == []:
            return None
        self.nums = nums
        return self.build(0, len(nums))


if __name__ == "__main__":
    solution = Solution()
    node = solution.sortedArrayToBST([1, 2, 3])
    print node.val, node.left.val, node.right.val

File: src/algorithms/cpp/Convert_Sorted_Array_to_Binary_Search_Tree.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    vector <int> nums;

private:
    TreeNode* build(int l, int r) {
        int mid = (l + r) >> 1;
        TreeNode *root = new TreeNode(nums[mid]);
        if (mid - l)
            root->left = build(l, mid);
        if (r - mid - 1)
            root->right = build(mid + 1, r);
        return root;
    }

public:
    TreeNode* sortedArrayToBST(vector <int> &nums) {
        if (SZ(nums) == 0)      return NULL;
        this->nums = nums;
        return build(0, SZ(nums));
    }
};

int main() {
    return 0;
}
